txsni/snimap.py

```diff
@@ -112,25 +153,48 @@ def __setattr__(self, attr, val):
     def __delattr__(self, attr):
         return delattr(self._obj, attr)
 
-
 @implementer(IOpenSSLServerConnectionCreator)
 class SNIMap(object):
-    def __init__(self, mapping):
+    def __init__(self, mapping, acme_mapping=None):
         self.mapping = mapping
+        self.acme_mapping = acme_mapping
         self._negotiationDataForContext = collections.defaultdict(
             _NegotiationData
         )
         try:
-            self.context = self.mapping['DEFAULT'].getContext()
+            self.context = self.mapping[b'DEFAULT'].getContext()
         except KeyError:
             self.context = CertificateOptions().getContext()
         self.context.set_tlsext_servername_callback(
             self.selectContext
         )
+        self.counter = 0
+
+    def selectAlpn(self, default, connection, protocols):
+        """
+        Trap alpn negotation.
+
+        Acme works by sending a special certificate based on this negotiation.
+        If such a certificate exists in self.acme_mapping, we will respond.
+
+        OpenSSL design flaws prevent us from changing certificates here.
+        Instead we detect acme earlier and send the certificate in
+        selectContext()
+        """
+        if (not ACME_TLS_1 in protocols
+            or not self.acme_mapping
+            or not connection._acme_tls_1):
+            return default()
+
+        return ACME_TLS_1
 
     def selectContext(self, connection):
+        mapping = self.mapping
+        if connection._acme_tls_1 and self.acme_mapping:
+            mapping = self.acme_mapping
```

Comment:
I thought about catching KeyError from self.acme_mapping to fall back on self.mapping, but if you somehow managed to own acme-tls/1.com it's your own fault.